File: backend/src/routes/sessions.js
import { Router } from "express";
import multer from "multer";

import { ChatSession, Message, LoanProduct } from "../models/index.js";
import { SenderType } from "../constants.js";
import { HttpError } from "../errors.js";
import {
  asyncHandler,
  authenticate,
  getUserSessionOr404,
  validateBody,
} from "../middleware.js";
import { sessionCreateSchema, messageCreateSchema } from "../schemas.js";
import { serializeMessage, serializeSession } from "../serializers.js";
import {
  generateAssistantReply,
  generateConsultationSummary,
} from "../services/chatService.js";
import { extractTextFromPdf } from "../services/pdfService.js";
import {
  advanceStateAfterDocument,
  advanceStateAfterMessage,
  advanceStateAfterSummary,
  initializeSessionState,
} from "../services/sessionStateService.js";

const upload = multer({ storage: multer.memoryStorage() });

export const sessionsRouter = Router();
sessionsRouter.use(authenticate);

sessionsRouter.post(
  "/",
  validateBody(sessionCreateSchema),
  asyncHandler(async (req, res) => {
    const payload = req.validated;
    const chatSession = ChatSession.build({
      user_id: req.user.id,
      title: payload.title || `${payload.loan_purpose} advice`,
      intent: payload.intent,
      loan_amount: payload.loan_amount,
      loan_purpose: payload.loan_purpose,
      monthly_income: payload.monthly_income,
      employment_type: payload.employment_type,
      existing_monthly_emi: payload.existing_monthly_emi,
      preferred_tenure_months: payload.preferred_tenure_months,
      risk_profile: payload.risk_profile,
      state_snapshot: {},
    });
    initializeSessionState(chatSession);
    await chatSession.save();
    res.status(201).json(serializeSession(chatSession));
  }),
);

sessionsRouter.get(
  "/",
  asyncHandler(async (req, res) => {
    const sessions = await ChatSession.findAll({
      where: { user_id: req.user.id },
      order: [["updated_at", "DESC"]],
    });
    res.json(sessions.map(serializeSession));
  }),
);

sessionsRouter.get(
  "/:sessionId",
  asyncHandler(async (req, res) => {
    const chatSession = await getUserSessionOr404(
      req.params.sessionId,
      req.user.id,
    );
    res.json(serializeSession(chatSession));
  }),
);

sessionsRouter.get(
  "/:sessionId/messages",
  asyncHandler(async (req, res) => {
    await getUserSessionOr404(req.params.sessionId, req.user.id);
    const messages = await Message.findAll({
      where: { session_id: req.params.sessionId },
      order: [["created_at", "ASC"]],
    });
    res.json(messages.map(serializeMessage));
  }),
);

sessionsRouter.post(
  "/:sessionId/messages",
  validateBody(messageCreateSchema),
  asyncHandler(async (req, res) => {
    const { sessionId } = req.params;
    const chatSession = await getUserSessionOr404(sessionId, req.user.id);

    const userMessage = await Message.create({
      session_id: sessionId,
      sender_type: SenderType.USER,
      message: req.validated.message,
      metadata: { source: "chat-ui" },
    });

    const products = await LoanProduct.findAll({
      order: [["interest_rate", "ASC"]],
    });
    const history = await Message.findAll({
      where: { session_id: sessionId },
      order: [["created_at", "ASC"]],
    });

    const assistantReply = await generateAssistantReply({
      chatSession,
      userMessage: req.validated.message,
      products,
      messageHistory: history,
    });

    const assistantMessage = await Message.create({
      session_id: sessionId,
      sender_type: SenderType.ASSISTANT,
      message: assistantReply.answer,
      metadata: {
        ...assistantReply.metadata,
        disclaimer: assistantReply.disclaimer,
      },
    });

    advanceStateAfterMessage({
      chatSession,
      userMessage: req.validated.message,
      assistantMessage: assistantReply.answer,
      toolOutputs: assistantReply.tool_outputs,
    });
    chatSession.updated_at = new Date();
    await chatSession.save();

    res.json({
      session: serializeSession(chatSession),
      user_message: serializeMessage(userMessage),
      assistant_message: serializeMessage(assistantMessage),
      eligible_products: assistantReply.eligible_products,
      tool_outputs: assistantReply.tool_outputs,
      disclaimer: assistantReply.disclaimer,
    });
  }),
);

sessionsRouter.post(
  "/:sessionId/upload",
  upload.single("file"),
  asyncHandler(async (req, res) => {
    const { sessionId } = req.params;
    const file = req.file;
    const allowedTypes = new Set([
      "application/pdf",
      "application/octet-stream",
    ]);
    if (!file || !allowedTypes.has(file.mimetype)) {
      throw new HttpError(400, "Only PDF uploads are supported");
    }

    const chatSession = await getUserSessionOr404(sessionId, req.user.id);
    const extractedText = await extractTextFromPdf(file.buffer);
    if (!extractedText.trim()) {
      throw new HttpError(400, "Could not extract text from the uploaded PDF");
    }

    const filename = file.originalname || "uploaded.pdf";
    const summary = await generateConsultationSummary(
      "Create a concise document summary for this borrower submission.\n\n" +
        `Document text:\n${extractedText.slice(0, 12000)}`,
      { stateSnapshot: chatSession.state_snapshot || {}, summaryKind: "document" },
    );

    const documentMessage = await Message.create({
      session_id: sessionId,
      sender_type: SenderType.SYSTEM,
      message: summary,
      metadata: {
        tool_called: "pdf_extractor",
        filename,
        extracted_characters: extractedText.length,
      },
    });

    advanceStateAfterDocument({ chatSession, filename });
    chatSession.updated_at = new Date();
    if (!chatSession.summary) {
      chatSession.summary = summary;
    }
    await chatSession.save();

    res.json({
      session_id: sessionId,
      filename,
      extracted_text_preview: extractedText.slice(0, 500),
      summary,
      message: serializeMessage(documentMessage),
    });
  }),
);

sessionsRouter.post(
  "/:sessionId/summary",
  asyncHandler(async (req, res) => {
    const { sessionId } = req.params;
    const chatSession = await getUserSessionOr404(sessionId, req.user.id);
    const messages = await Message.findAll({
      where: { session_id: sessionId },
      order: [["created_at", "ASC"]],
    });
    if (messages.length === 0) {
      throw new HttpError(400, "Cannot summarize an empty conversation");
    }

    const conversationText = messages
      .map((message) => `${message.sender_type}: ${message.message}`)
      .join("\n");
    const summary = await generateConsultationSummary(conversationText, {
      stateSnapshot: chatSession.state_snapshot || {},
      summaryKind: "conversation",
    });

    advanceStateAfterSummary(chatSession);
    chatSession.summary = summary;
    chatSession.updated_at = new Date();
    await chatSession.save();

    res.json({ session_id: sessionId, summary });
  }),
);

File: backend/app/api/routes/sessions.py
from datetime import datetime

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db, get_user_session_or_404
from app.models.chat import ChatSession, Message, SenderType
from app.models.loan_product import LoanProduct
from app.models.user import User
from app.schemas.message import (
    ChatReplyResponse,
    MessageCreate,
    MessageResponse,
    SummaryResponse,
    UploadResponse,
)
from app.schemas.session import SessionCreate, SessionResponse
from app.services.chat_service import generate_assistant_reply, generate_consultation_summary
from app.services.pdf_service import extract_text_from_pdf
from app.services.session_state_service import (
    advance_state_after_document,
    advance_state_after_message,
    advance_state_after_summary,
    initialize_session_state,
)

router = APIRouter(tags=["sessions"])


def serialize_message(message: Message) -> MessageResponse:
    return MessageResponse(
        id=message.id,
        session_id=message.session_id,
        sender_type=message.sender_type,
        message=message.message,
        metadata=message.meta or {},
        created_at=message.created_at,
    )


def serialize_session(chat_session: ChatSession) -> SessionResponse:
    return SessionResponse.model_validate(chat_session)


@router.post("/sessions", response_model=SessionResponse, status_code=status.HTTP_201_CREATED)
def create_session(
    payload: SessionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SessionResponse:
    chat_session = ChatSession(
        user_id=current_user.id,
        title=payload.title or f"{payload.loan_purpose} advice",
        intent=payload.intent,
        loan_amount=payload.loan_amount,
        loan_purpose=payload.loan_purpose,
        monthly_income=payload.monthly_income,
        employment_type=payload.employment_type,
        existing_monthly_emi=payload.existing_monthly_emi,
        preferred_tenure_months=payload.preferred_tenure_months,
        risk_profile=payload.risk_profile,
        state_snapshot={},
    )
    initialize_session_state(chat_session)
    db.add(chat_session)
    db.commit()
    db.refresh(chat_session)
    return serialize_session(chat_session)


@router.get("/sessions", response_model=list[SessionResponse])
def list_sessions(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[SessionResponse]:
    sessions = (
        db.query(ChatSession)
        .filter(ChatSession.user_id == current_user.id)
        .order_by(ChatSession.updated_at.desc())
        .all()
    )
    return [serialize_session(item) for item in sessions]


@router.get("/sessions/{session_id}", response_model=SessionResponse)
def get_session(
    session_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SessionResponse:
    chat_session = get_user_session_or_404(session_id, db, current_user.id)
    return serialize_session(chat_session)


@router.get("/sessions/{session_id}/messages", response_model=list[MessageResponse])
def list_messages(
    session_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[MessageResponse]:
    get_user_session_or_404(session_id, db, current_user.id)
    messages = (
        db.query(Message)
        .filter(Message.session_id == session_id)
        .order_by(Message.created_at.asc())
        .all()
    )
    return [serialize_message(message) for message in messages]


@router.post("/sessions/{session_id}/messages", response_model=ChatReplyResponse)
async def create_message(
    session_id: str,
    payload: MessageCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ChatReplyResponse:
    chat_session = get_user_session_or_404(session_id, db, current_user.id)

    user_message = Message(
        session_id=session_id,
        sender_type=SenderType.USER,
        message=payload.message,
        meta={"source": "chat-ui"},
    )
    db.add(user_message)
    db.commit()
    db.refresh(user_message)

    products = db.query(LoanProduct).order_by(LoanProduct.interest_rate.asc()).all()
    history = (
        db.query(Message)
        .filter(Message.session_id == session_id)
        .order_by(Message.created_at.asc())
        .all()
    )
    assistant_reply = await generate_assistant_reply(
        chat_session=chat_session,
        user_message=payload.message,
        products=products,
        message_history=history,
    )

    assistant_message = Message(
        session_id=session_id,
        sender_type=SenderType.ASSISTANT,
        message=assistant_reply.answer,
        meta={**assistant_reply.metadata, "disclaimer": assistant_reply.disclaimer},
    )
    db.add(assistant_message)
    advance_state_after_message(
        chat_session=chat_session,
        user_message=payload.message,
        assistant_message=assistant_reply.answer,
        tool_outputs=assistant_reply.tool_outputs,
    )
    chat_session.updated_at = datetime.utcnow()
    db.commit()
    db.refresh(assistant_message)
    db.refresh(chat_session)

    return ChatReplyResponse(
        session=serialize_session(chat_session),
        user_message=serialize_message(user_message),
        assistant_message=serialize_message(assistant_message),
        eligible_products=assistant_reply.eligible_products,
        tool_outputs=assistant_reply.tool_outputs,
        disclaimer=assistant_reply.disclaimer,
    )


@router.post("/sessions/{session_id}/upload", response_model=UploadResponse)
async def upload_pdf(
    session_id: str,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> UploadResponse:
    if file.content_type not in {"application/pdf", "application/octet-stream"}:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Only PDF uploads are supported",
        )

    chat_session = get_user_session_or_404(session_id, db, current_user.id)
    extracted_text = extract_text_from_pdf(await file.read())
    if not extracted_text.strip():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Could not extract text from the uploaded PDF",
        )

    summary = await generate_consultation_summary(
        messages_text=(
            "Create a concise document summary for this borrower submission.\n\n"
            f"Document text:\n{extracted_text[:12000]}"
        ),
        state_snapshot=chat_session.state_snapshot or {},
        summary_kind="document",
    )

    document_message = Message(
        session_id=session_id,
        sender_type=SenderType.SYSTEM,
        message=summary,
        meta={
            "tool_called": "pdf_extractor",
            "filename": file.filename,
            "extracted_characters": len(extracted_text),
        },
    )
    db.add(document_message)
    advance_state_after_document(
        chat_session=chat_session,
        filename=file.filename or "uploaded.pdf",
    )
    chat_session.updated_at = datetime.utcnow()
    if not chat_session.summary:
        chat_session.summary = summary
    db.commit()
    db.refresh(document_message)

    return UploadResponse(
        session_id=session_id,
        filename=file.filename or "uploaded.pdf",
        extracted_text_preview=extracted_text[:500],
        summary=summary,
        message=serialize_message(document_message),
    )


@router.post("/sessions/{session_id}/summary", response_model=SummaryResponse)
async def summarize_session(
    session_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SummaryResponse:
    chat_session = get_user_session_or_404(session_id, db, current_user.id)
    messages = (
        db.query(Message)
        .filter(Message.session_id == session_id)
        .order_by(Message.created_at.asc())
        .all()
    )
    if not messages:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot summarize an empty conversation",
        )

    conversation_text = "\n".join(
        f"{message.sender_type.value}: {message.message}" for message in messages
    )
    summary = await generate_consultation_summary(
        conversation_text,
        state_snapshot=chat_session.state_snapshot or {},
        summary_kind="conversation",
    )
    advance_state_after_summary(chat_session)
    chat_session.summary = summary
    chat_session.updated_at = datetime.utcnow()
    db.commit()
    db.refresh(chat_session)

    return SummaryResponse(session_id=session_id, summary=summary)
